### DISCUSSION TITLE
Why is my trained transformer model(en_core_web_trf) ineffective?

### DISCUSSION THREAD

**Discussion by qingyun1988** [explosion/spaCy#7942] [Help: Other Questions] [answered]
I used flowing code to train the 'en_core_web_trf' model, but the consequence seems to be ineffective.

```
TRAIN_DATA = [
    (
        "Books were present at the meeting.",
        {
            "heads": [1,1,1,2,5,3,1],
            "deps": ["nsubj", "ROOT", "acomp", "prep", "det", "pobj","punct"],
        }
    ),
    (
        "They were present at the meeting.",
        {
            "heads": [1,1,1,2,5,3,1],
            "deps": ["nsubj", "ROOT", "acomp", "prep", "det", "pobj","punct"],
        },
    ),
    (
        "Aides were present at the meeting.",
        {
            "heads": [1,1,1,2,5,3,1],
            "deps": ["nsubj", "ROOT", "acomp", "prep", "det", "pobj","punct"],
        },
    ),
]

@plac.annotations(
    model=("Model name. Defaults to blank 'en' model.", "option", "m", str),
    output_dir=("Optional output directory", "option", "o", Path),
    n_iter=("Number of training iterations", "option", "n", int),
)
def main(model="en_core_web_trf", output_dir="D:\Python\spacy\customTrf30", n_iter=20):#en_core_web_trf
    """Load the model, set up the pipeline and train the parser."""
    nlp = spacy.load(model)
    parser = nlp.get_pipe("parser")

    # add labels to the parser
    for _, annotations in TRAIN_DATA:
        for dep in annotations.get("deps", []):
            parser.add_label(dep)

    # get names of other pipes to disable them during training
    #pipe_exceptions = ["parser","tok2vec"]#for web_lg
    #pipe_exceptions = ["parser","transformer"]#for web_trf
    pipe_exceptions = ["parser","transformer","tok2vec"]
    other_pipes = [pipe for pipe in nlp.pipe_names if pipe not in pipe_exceptions]
    with nlp.disable_pipes(*other_pipes):  # only train parser
        optimizer = nlp.resume_training()#nlp.begin_training()//resume_training
        for itn in range(n_iter):
            random.shuffle(TRAIN_DATA)
            losses = {}
            for text, annotations in TRAIN_DATA:
                example = Example.from_dict(nlp.make_doc(text), annotations)
                #losses = nlp.rehearse([example],sgd=optimizer)#,component_cfg=pipe_exceptions
                nlp.update([example],sgd=optimizer, losses=losses)

            print("Losses", losses)

    # save model to output directory
    if output_dir is not None:
        output_dir = Path(output_dir)
        if not output_dir.exists():
            output_dir.mkdir()
        nlp.to_disk(output_dir)
        print("Saved model to", output_dir)
```

However, if I use the same codes to the 'en_core_web_lg' model, it seems to work fine and by testing it can show immediately the consequence I expect.

The difference is only to replace 'en_core_web_trf' with 'en_core_web_lg'.

Is there any other way to train the 'en_core_web_trf' model? How can I train transformer model(en_core_web_trf)?

In addition, I don't know how to train models with training config file. 

Appreciate if anyone can help. Thank you!

**Comment by polm** [comment, marked as answer]
Sorry to hear you're having trouble. Can you clarify a few things?

1. Are those three examples your entire training data?
2. When you say it works with the large model but not Transformers, what changes are you referring to? What do you check to see it changed?
3. What are you trying to change in the model? Those sentences look pretty normal.
4. Did you see the [training quickstart](https://spacy.io/usage/training#quickstart)? That will show you how to train a Transformer model with a config.

If you actually only have three training examples, that's not enough to train any of the models meaningfully.

**Reply by qingyun1988** [reply]
Thank you for your reply.

Yes, those three examples are entire training data, and I know they are not enough to train a model.

I needn't train a new model from scratch. I am using the pre-trained model 'en_core_web_trf' for my own app, but it doesn't perform well on some sentences, e.g. those 3 examples above.

So I want to improve or correct these performances for these sentences in the model 'en_core_web_trf'.  This is the reason why there are only 3 sentences in the entire training data.

Of course, there are more sentences in future. This will depend on what particular sentences I'll encounter in future perform incorrect that is out of my expectation.

The changes I am referring to is the expected correctness of the dependencies of these sentences testing in the new model I've just trained.

Yes, these sentences above are pretty normal. However, the results of parsing those sentences for their dependencies are different. The results performing in the large model is more expected for me but the transformer isn't.

So, I deliberately gave some incorrect dependencies to the large model for training data.
`        "They were present at the meeting.",
        {
            "heads": [1,1,1,1,5,3,1],
            "deps": ["nsubj", "ROOT", "acomp", "prep", "det", "pobj","punct"],
        }`

After training it, I display it using 'spacy.displacy', and the results are as I expected.

But the transformer model can't perform as same as the large model. 
`        "They were present at the meeting.",
        {
            "heads": [1,1,1,2,5,3,1],
            "deps": ["nsubj", "ROOT", "acomp", "prep", "det", "pobj","punct"],
        }`

The displayed dependencies are the same as the one performing in the original model as if the model hasn't trained.

So, I suspect I've done something wrong?

In addition, I have seen the training quickstart. but I don't know how to provide training data for training model with a config. I haven't seen the command to provide training data.

My English is not good. So, please forgive me for my lack of meaning.

**Reply by qingyun1988** [reply]
Does it have to be used for the config training method to work?

**Reply by qingyun1988** [reply]
if a few examples are not enough to train the model, why are they enough for the large model but not the transformer model? I've found that only an example is enough to train for the large model since I have used Spacy. I mean the results for the current sentence not another one in the same circumstance.

**Reply by honnibal** [reply]
Hi @qingyun1988 ,

We recommend the training config (plus CLI) or at least the more general training methods because there's a lot of details to get right in the training loop. In this case, I think the problem is probably that you're not stepping the learning rate schedule. This would mean that the learning rate is stuck at 0, which means that the optimization is not meaningfully changing the weights. It might actually leave the weights identical -- I'm not sure whether there's a regularization that is applied regardless of the learning rate.

You could try the `train_while_improving` function from `spacy.training.loop` instead of your own training loop code. It works as a generator, so you still get a block to execute your own code on each batch. If you do want to use your own code, some things to check are the learning rate and whether the weights of the models are actually changing. You can check the weights by doing:

```python
for name, pipe in nlp.pipeline:
    for node in pipe.model.walk():
        for param_name in node.param_names:
            param = node.get_param(param_name) # (numpy or cupy ndarray)
```

Unfortunately, even if your code is entirely correct, the transformers are not guaranteed to converge on a 3 sentence sample. The weights should definitely change, though -- so you can at least exclude some problems. The reason it might not converge is batch size: the optimization isn't stable at small batches with the default learning rate. You can fiddle with the learning rate, or use a slightly larger set of example sentences to memorize, something like 10 or 30 should be fine.

**Reply by qingyun1988** [reply]
Hi，@honnibal!
I'm so sorry to have just seen your reply, thanks for your patient and kindness.

I am using `training API` but not `training config` method because it's convenient to using dependency editor which was developed by myself to create gold-standard annotations. 

And I don't know how to use `training config` method to train a already existed model which seems to be harder than the former for me.  What are the problems for keeping to use  `training API` method?

How to try` train_while_improving` function from `spacy.training.loop`?  Does `nlp.resume_training() `function of `training API` 
 equate ` train_while_improving` function?

What is the `more general training methods`? 

By the way, what is the best way to get or create gold-standard annotations? My solution was to develop a dependency editor.

Thank you!

**Reply by qingyun1988** [reply]
@honnibal
